clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -396,46 +450,31 @@ synchronized void resetProducerId() {
             throw new IllegalStateException("Cannot reset producer state for a transactional producer. " +
                     "You must either abort the ongoing transaction or reinitialize the transactional producer instead");
         setProducerIdAndEpoch(ProducerIdAndEpoch.NONE);
-        this.nextSequence.clear();
-        this.lastAckedSequence.clear();
-        this.inflightBatchesBySequence.clear();
+        topicPartitionBookeeper.reset();
         this.partitionsWithUnresolvedSequences.clear();
-        this.lastAckedOffset.clear();
     }
 
     /**
      * Returns the next sequence number to be written to the given TopicPartition.
      */
     synchronized Integer sequenceNumber(TopicPartition topicPartition) {
-        Integer currentSequenceNumber = nextSequence.get(topicPartition);
-        if (currentSequenceNumber == null) {
-            currentSequenceNumber = 0;
-            nextSequence.put(topicPartition, currentSequenceNumber);
-        }
-        return currentSequenceNumber;
+        if (!topicPartitionBookeeper.hasSequenceNumber(topicPartition))
+            topicPartitionBookeeper.initSequenceNumber(topicPartition);
+
+        return topicPartitionBookeeper.get(topicPartition).nextSequenceNumber;
     }
 
     synchronized void incrementSequenceNumber(TopicPartition topicPartition, int increment) {
-        Integer currentSequenceNumber = nextSequence.get(topicPartition);
-        if (currentSequenceNumber == null)
-            throw new IllegalStateException("Attempt to increment sequence number for a partition with no current sequence.");
```

Comment:
The check here is not needed as there is only one usage of this method which is right after `sequenceNumber()` in `RecordAccumulator.drainBatchesForOneNode`. Removed the corresponding test too.